feat: add prettier as formatter (#1030)
CleverCloud/clever-components: docs/adr/adr-0026-introducing-prettier.md

## Solution

### Why did we choose Prettier?

We decided to go for Prettier for various reasons:

- It is widely used and maintained.
- Has integration with IDEs.
- It enforces code style. Other than our internal conventions we didn't have a proper formatter enforcing.

Note: As we haven't updated ESLint yet, see details on the dedicated section below, we've added the Prettier ESLint plugin so that ESLint formatting rules doesn't conflict with Prettier.

### Why didn't we use ESLint Stylistic?

At some point we considered going for the `stylistic` package for ESLint to keep our usage.
While it would have been great, this wouldn't solve the lack of formatter for Stylelint.

Considering this, we tried to use the `stylelint-prettier` plugin to format only our CSS.
However, we couldn't do that because:
`postcss-lit` uses babel and if you use that parser `stylelint-prettier` will just skip your file.
So we tried to force Prettier to use the `css` parser instead for Stylelint but that resulted in some errors.

Knowing that we couldn't use this plugin and thus not having a formatter for our Stylelint/CSS part we decided not to use Stylistic.

> Moreover, if we wanted to use both ESLint with Stylystic and prettier for the CSS it would have caused some conflicts.

Reviewer: **🔨 issue:**

```suggestion
Moreover, if we wanted to use both ESLint with Stylistic and prettier for the CSS it would have caused some conflicts.
```

kind: '📌 Architecture Decision Records'
---
# ADR 0026: Introducing Prettier as a formatter

🗓️ 2024-06-13 · ✍️ Mathieu Degand

> This ADR tries to explain why we decided to introduce Prettier while migrating our linting tools (ESLint/Stylelint) to their latest version.

Reviewer 2: ```suggestion
This ADR explains why we decided to introduce Prettier while migrating our linting tools (ESLint/Stylelint) to their latest version.
```

## Context

> One year ago we made the decision to add Stylelint to our components library. You can find the how and why on the [ADR]().

Reviewer 2: question: is there a missing link here `[ADR]()` ?

> One year ago we made the decision to add Stylelint to our components library. You can find the how and why on the [ADR]().

Reviewer 3: **nitpick:** 
```suggestion
One year ago we made the decision to add Stylelint to our components library. You can find the how and why in the [ADR]().
```

At that time Stylelint was in version `14`ish and is now in version `16` as of this writing.
During a meeting while discussing versions of the tools we use, we decided to take a look at the updates and the feasibility to update them to the latest version.
And here came the problem: linters don't (want to) format anymore.

That's the case for Stylelint, that decided to deprecate format rules in version `15`, while removing them all in the latest version (16).
But that's also the case for ESLint that made the same decision.
In fact most if not all of the linters tend to switch to not format the code now.

> Thus said we took the decision to handle the subject on our side and find a formatter before updating both tools. (ESLint and Stylelint)

Reviewer 3: **nitpick:**
```suggestion
Thus said, we took the decision to handle the subject on our side and find a formatter before updating both ESLint and Stylelint.
```

Another reason was that while we had some basic formatting conventions, it wasn't relying on any tool except our IDEs which could lead to inconsistencies in our codebase.

## Formatters study

Before choosing a formatter we had to test some in order to find the one that would suit us best.
You'll find below the ones that we tried and some explanations....

### Biome

Biome is a toolchain written in Rust that can both format and lint but also more. See [Biome's website](https://biomejs.dev/) for more information.

#### Pros

- Biome is multipurpose, it has a formatter, a linter and more.

#### Cons

- It is opiniated, which means you can't really override things to match your need.
- It only has some default agnostic formatting options.
- While being multipurpose is a pro, it's also a con. If you want to use only one feature, you'll still need to get the whole package and deactivate the features you don't need.

> While we could have chosen to go for Biome as a formatter, and maybe go even further and choose it as a linter too, there's a few reasons while we decided not to go for this solution.

Reviewer 3: **nitpick:**
```suggestion
While we could have chosen to go for Biome as a formatter, and maybe go even further and choose it as a linter too, there are a few reasons while we decided not to go for this solution.
```

> - As a linter, there's not a lot of options and it can't check our CSS, so that's a clear no go.

Reviewer 2: nitpick can you add that it is because our CSS is inside JS files

- If we wanted to only go for the formatter, we still had to get everything and deactivate features we didn't need/want. (formatter, analyzer..)
- Biome has the same philosophy as Prettier for the formatting but Prettier is more known and used for now.

### dprint

From [dprint's website](https://dprint.dev/):

> A pluggable and configurable code formatting platform written in Rust.

#### Pros

- Lot of options.
- You can develop plugins.
- If you want to develop plugins, there's a rust SDK.

#### Cons

- Even though it's possible to write plugins in another language, resources about this are sparse. It means you better be proficient with Rust.

### Prettier

[Prettier](https://prettier.io) is an opiniated code formatter.

#### Pros

- Can format both JS, CSS, template literals in one take.

#### Cons

> - There's not a lot of options.

Reviewer 3: **nitpick:**
```suggestion
- There are not a lot of options.
```

> - There's not a lot of options.
>   - It is justified by the `opiniated formatter` logic.

Reviewer 2: nitpick: Maybe one tick is enough.

```suggestion
- There's not a lot of options which is justified by the `opiniated formatter` logic.
```

> Prettier is the option we chose to use, you can find more information about the why on the dedicated section below.

Reviewer 3: **nitpick:**
```suggestion
Prettier is the option we chose to use, you can find more information about the why within the dedicated section below.
```

### js-beautify

#### Pros

- Has integrations with IDEs

#### Cons

- Unsure about the future of the maintenance ([See github README](https://github.com/beautifier/js-beautify?tab=readme-ov-file#contributors-needed))
- Can over-format

### Others

- While we knew all these tools existed we wanted to know what Lit users were using
  - Knowing that we asked the Lit discord's what they were using
    - Someone told us that they were using the new `Stylistic` package for ESLint that keeps formatting rules
    - They were also using a Prettier plugin for CSS
    - While it seemed a good idea at first, you'll find why it wasn't adapted to our case below.

> ## Investigation after Lit Discord's question (TODO: remove this)

Reviewer 3: **nitpick:** You have a TODO here :eyes:

> - I have asked the Lit Discord's to know if some users used specific tools to format their Lit components

Reviewer 3: **nitpick:** I would stick with "We" since it's what we use everywhere and I don't think we ever go for "I" within ADRs but it's not very important so you decide of course :wink: 
```suggestion
- We have asked the Lit Discord to know if some users used specific tools to format their Lit components
```
or
```suggestion
- We have asked the Lit's Discord to know if some users used specific tools to format their Lit components
```

>   - It seems like they're using a `Stylistic` package that should maintains the deprecated rules

Reviewer 3: **nitpick:**
```suggestion
  - It seems like they're using a `Stylistic` package that should maintain the deprecated rules
```

  - However, as the new ESLint version is quite new (04-05-2024) there are a lot of packages which are not working on version 9
    - I suggest that we don't migrate to version 9 until some of the plugins are updated or that we find another formatter
  - About Stylelint:
    - Version 15 deprecated its formatting rules, while removing them from version 16
    - Processors have been removed and changed to `customSyntax`
      - You can have multiple of them for different purposes
        - e.g: `postcss-lit` for `.js` component files
        - e.g: `postcss` for `.css` files
      - This introduction made the Stylelint `postcss-css-in-js` processor/plugin obsolete.
        - Luckily there's now a `customSyntax` for

## Solution

### Why did we choose Prettier?

We decided to go for Prettier for various reasons:

- It is widely used and maintained.
- Has integration with IDEs.

> - It enforces code style. Beside our internal conventions we didn't have a proper formatter enforcing.

Reviewer 3: **nitpick:** 
```suggestion
- It enforces code style. Other than our internal conventions we didn't have a proper formatter enforcing.
```

> Note: As we didn't update ESLint yet, see details on the dedicated section below, we added the Prettier ESLint plugin so that ESLint formatting rules doesn't conflict with Prettier.

Reviewer 3: **nitpick:**
```suggestion
Note: As we haven't updated ESLint yet, see details on the dedicated section below, we've added the Prettier ESLint plugin so that ESLint formatting rules doesn't conflict with Prettier.
```

`postcss-lit` uses babel and if you use that parser `stylelint-prettier` will just skip your file.
So we tried to force Prettier to use the `css` parser instead for Stylelint but that resulted in some errors.

Knowing that we couldn't use this plugin and thus not having a formatter for our Stylelint/CSS part we decided not to use Stylistic.
Moreover, if we wanted to use both ESLint with Stylystic and prettier for the CSS it would have caused some conflicts.

### Update of Stylelint

- Like ESLint, Stylelint version 15 got rid of its formatting rules.
- Now, that Prettier handles the formatting we considered updating Stylelint to its latest version.

For this we did:

- Update the Stylelint version.
- Got rid of the `postcss-css-in-js` custom syntax as it's now deprecated.
  - Decided to use the dedicated `postcss-lit` syntax instead.
- Removed our idiomatic CSS properties order.
  - We took this decision because the package was/will not be maintain(ed).
  - We chose to use the `stylelint-order` package instead and use an alphabetical order.
- Removed all the Stylelint formatting rules from our configuration file.

### Why didn't we update ESLint (for now)?

> With the update of Stylelint and the introduction of Prettier, we also considered to update ESLint to version 9. However we decided after some tests to not do it yet.

Reviewer 3: **nitpick:**
```suggestion
With the update of Stylelint and the introduction of Prettier, we also considered to updating ESLint to version 9. However, after some testing, we have decided to hold it off for now.
```
